crates/ark/src/lsp/completions/sources/composite.rs

```diff
@@ -78,16 +86,64 @@ pub(crate) fn get_completions(
         )?;
     }
 
-    // Remove duplicates
-    let mut uniques = HashSet::new();
-    completions.retain(|x| uniques.insert(x.label.clone()));
+    // Simplify to plain old CompletionItems and sort them
+    let completions = finalize_completions(completions);
+
+    Ok(Some(completions))
+}
 
-    // Sort completions by providing custom 'sort' text to be used when
-    // ordering completion results. we use some placeholders at the front
-    // to 'bin' different completion types differently; e.g. we place parameter
-    // completions at the front, followed by variable completions (like pipe
-    // completions and subset completions), followed by anything else.
-    for item in &mut completions {
+fn push_completions<S>(
+    source: S,
+    completion_context: &CompletionContext,
+    completions: &mut HashMap<String, CompletionItemWithSource>,
+) -> anyhow::Result<()>
+where
+    S: CompletionSource,
+{
+    let source_name = source.name();
+
+    if let Some(source_completions) = collect_completions(source, completion_context)? {
+        for item in source_completions {
+            if let Some(existing) = completions.get(&item.label) {
+                log::debug!(
```

Comment:
```suggestion
                log::trace!(
```

We almost unconditionally skip straight to trace! level

Reply:
Sorry this keeps happening intermittently! I do know about the preference for `log::trace!`. Sometimes it's me and sometimes I think it's co-pilot (and me not catching it).